tests/src/test/java/alluxio/collections/IndexedSetIntegrationTest.java

```diff
@@ -0,0 +1,259 @@
+    @Override
+    public void run() {
+      while (!mStopThreads.get()) {
+        Iterator<TestInfo> it = mIndexedSet.iterator();
+        while (it.hasNext()) {
+          it.next();
+          it.remove();
+          mCount++;
+        }
+      }
+    }
+  }
+
+  private class ConcurrentClear extends ConcurrentTask {
+    @Override
+    public void run() {
+      while (!mStopThreads.get()) {
+        mIndexedSet.clear();
+        mCount++;
+      }
+    }
+  }
+
+  private final class TestInfo {
+    private long mId;
+    private int mSize;
+
+    private TestInfo() {
+      mId = java.util.concurrent.ThreadLocalRandom.current().nextLong();
+      mSize = java.util.concurrent.ThreadLocalRandom.current().nextInt(0, MAX_SIZE);
+    }
+
+    public long getId() {
+      return mId;
+    }
+
+    public int getSize() {
+      return mSize;
+    }
+  }
+
+  private final IndexedSet.FieldIndex<TestInfo> mIdIndex =
+      new IndexedSet.FieldIndex<TestInfo>() {
+        @Override
+        public Object getFieldValue(TestInfo o) {
+          return o.getId();
+        }
+      };
+
+  private final IndexedSet.FieldIndex<TestInfo> mSizeIndex =
+      new IndexedSet.FieldIndex<TestInfo>() {
+        @Override
+        public Object getFieldValue(TestInfo o) {
+          return o.getSize();
+        }
+      };
+
+  @Before
+  public void before() throws Exception {
+    mIndexedSet = new IndexedSet<>(mIdIndex, mSizeIndex);
+    mThreadPool = Executors.newCachedThreadPool();
+    mStopThreads = new AtomicBoolean(false);
+  }
+
+  @After
+  public void after() {
+    mThreadPool.shutdownNow();
+  }
+
+  /**
+   * Verifies the {@link #mIndexedSet} for internal consistency.
+   */
+  private void verifySet() {
+    Iterator<TestInfo> it = mIndexedSet.iterator();
+    Set<Long> ids = new HashSet<>();
+    Set<Integer> sizes = new HashSet<>();
+
+    // Verify the size.
+    int expectedCount = 0;
+    while (it.hasNext()) {
+      TestInfo info = it.next();
+      ids.add(info.getId());
+      sizes.add(info.getSize());
+      expectedCount++;
+    }
+    Assert.assertEquals(expectedCount, mIndexedSet.size());
+
+    // Verify the size according to the id index.
+    int count = 0;
+    for (Long id : ids) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mIdIndex, id);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+
+    // Verify the size according to the size index.
+    count = 0;
+    for (Integer size : sizes) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mSizeIndex, size);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+  }
+
+  @Test
+  public void basicConcurrentTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+    List<ConcurrentTask> addTasks = new ArrayList<>();
+    List<ConcurrentTask> removeTasks = new ArrayList<>();
+
+    addTasks.add(new ConcurrentAdd());
```

Comment:
it would be valuable to randomize the number of different tasks across different test executions